api/envoy/api/v2/route/route.proto

```diff
@@ -893,6 +893,20 @@ message RetryPolicy {
     google.protobuf.Duration max_interval = 2 [(validate.rules).duration = {gt {}}];
   }
 
+  message RetryBudget {
```

Comment:
@snowp could we fit this into the retry plugin concept (maybe with some generalization of the existing interface)? I have an interest in trying to decompose and shrink router/HCM, this could be a good place to test the waters.

Reply:
I actually thought about implementing this as a retry predicate, but opted for this way instead. It's pretty straightforward- should I go ahead and do this as a predicate?

Reply:
The current plugins don't currently allow aborting host selection: the predicate is only able to trigger another load balancing attempt. I can imagine generalizing the `RetryHostPredicate` interface from 

```
  virtual bool shouldSelectAnotherHost(const Host& candidate_host) PURE;
```

to

```
enum Decision {
  ACCEPTED,
  TRY_AGAIN,
  ABORT
};

  virtual Decision shouldSelectAnotherHost(const Host& candidate_host) PURE;

```

where ABORT makes the load balancer return nullptr + setting  a response flag.

Reply:
FWIW I think this is a pretty fundamental feature that we should probably support directly, but I don't feel that strongly about it as long as it is well documented, so will defer to all of you to decide on core vs. extension.

Reply:
Main concern I'd have with the retry plugin approach is that it relies on all the LB implementations checking the predicate. Right now only LoadBalancerBase LBs do, which means LBs such as maglev/ring_hash/original dst/etc would not honor it without doing some additional work there. There are also some potential issues around stats: using a retry plugin would still result in retry stats being incremented, but fail during load balancing. 

There might be a better potential extension point around RetryState: 

```
class RetryDecider {
  bool shouldRetry(const HeaderMap& request_headers, absl::optional<ResetReason> reset_reason, 
  const HeaderMap const * response_headers);
  
  // Checks whether the request/response satisfies the decider, without the implication of a
  // retry attempt actually being made. Used to decide whether a hedging attempt should be
  // accepted. 
  bool wouldRetry(const HeaderMap& request_headers, absl::optional<ResetReason> reset_reason, 
  const HeaderMap const * response_headers);
}
```
then expose something like

```
retry_requirements: 
- name: envoy.retry_decider.max_retry_count
   config:
      max_retries: 4 # rejects if retry count is exceeded
- name: envoy.retry_deciders.retry_on_any # retry if any of the following are true
  inner_deciders: 
    - name: envoy.retry_decider.retry_on_response_header
      config:
        header_matcher:
          name: :status
          exact_match: "500"
```

where a retry happens if all of the top level deciders return true, using a composite decider to implement "any of these could match".  This would decompose `RetryStateImpl` quite a bit and make it easy to implement the retry budget feature as just another decider, and opens up for making it easier to add retry policies that depend on maintaining some state, e.g. stop retrying 429s if you get more than X in a row. 

This is obviously a bunch more work, but might be a long term better solution as more and more retry behavior is added?

Reply:
I like that idea a lot, but I think it's too much to kick out for this PR. I'd be happy to write up the proposed proto/config changes in an issue and implement it all once everyone agrees on the interface.

Reply:
Yeah I also like this idea. Personally, I think we should just go ahead and add this feature directly and then circle back. IMO the retry code is pretty well abstracted away from the router so I don't really think there is that much risk. Thoughts?

Reply:
+1 to what @mattklein123 said. If @snowp and @htuch are on board with this, I'll go ahead and start drafting the issue and protos.

Reply:
I'm okay with implementing this directly and then aiming to implement the extension as a later improvement

Reply:
I don't want to block this PR, so do what makes most sense. I do have a general concern of the growth in what we consider fundamental features as previously discussed. If we can move to a model where we have more things outside the router/HCM, and we're able to opt them out of the build, this could be a win. That's really part of a broader discussion and not material to this PR.